### ISSUE TITLE
🧠 [PLAN] Decision Intelligence Transformation (ML/RL/Planner Integration)

### ISSUE THREAD

**Issue by dokuczacz** [dokuczacz/OmniFlowBeta#25]
# 🧠 [PLAN] Decision Intelligence Transformation (ML/RL/Planner Integration)

## 📋 Summary

This is a **6-week transformation plan** to migrate OmniFlow's decision-making from LLM-only architecture to a hybrid **ML + RL + Planner** system, where LLM is used only for text generation.

**Current State**: LLM decides intent → tool selection → response generation  
**Target State**: ML/RL decides intent/tools → LLM only generates text  

**Key Benefits**:
- ✅ Learn from 50+ historical decision episodes (imitation learning)
- ✅ Optimize via offline reinforcement learning (reward-based)
- ✅ Cost reduction: ~30% fewer LLM calls in production
- ✅ Performance: Faster decisions (ML inference < LLM API)
- ✅ No production risk: Offline training + shadow mode validation

---

## 📦 Documents (Merged to main)

All planning documents are in [`tmp/`](https://github.com/dokuczacz/OmniFlowBeta/tree/main/tmp):

1. **[AI_DECISION_BLUEPRINT.md](https://github.com/dokuczacz/OmniFlowBeta/blob/main/tmp/AI_DECISION_BLUEPRINT.md)** (30 KB)
   - Strategic overview: where is LLM today, where are decision points
   - Layer-by-layer mapping: LLM → ML/RL/Planner (9 components)
   - Lab ML architecture (directory structure, data flow)
   - 6-week timeline with gates & DoD
   - Real example: Task Management (PA-01) flow transformation
   - Shadow mode instrumentation guide
   - FAQ & risk mitigation

2. **[ML_INTEGRATION_POINTS.md](https://github.com/dokuczacz/OmniFlowBeta/blob/main/tmp/ML_INTEGRATION_POINTS.md)** (20 KB)
   - Concrete code locations (file:line references)
   - Before/after code snippets (copy-paste ready)
   - Integration patterns (flag-gating, fallback logic)
   - Environment variables to add
   - Testing & validation points
   - Rollout strategy (4 phases)

3. **[EXECPLAN_PHASE_1.md](https://github.com/dokuczacz/OmniFlowBeta/blob/main/tmp/EXECPLAN_PHASE_1.md)** (25 KB)
   - Phase 1: Policy Evaluation (ready to run today)
   - Setup instructions (venv, dependencies)
   - 3 Python scripts (snapshot explorer, metrics calculator, report generator)
   - Line-by-line commands
   - Expected outputs (CSV + JSON + HTML report)
   - Success criteria (DoD)

4. **[INDEX_AND_QUICK_START.md](https://github.com/dokuczacz/OmniFlowBeta/blob/main/tmp/INDEX_AND_QUICK_START.md)** (15 KB)
   - 2-minute executive summary
   - Quick navigation guide
   - Decision checklist (5 questions to approve)
   - Immediate next steps
   - Risk matrix & FAQ

---

## 🎯 The 6-Week Rollout Plan

| Week | Phase | Deliverable | Risk | Gate Criteria |
|------|-------|-------------|------|---------------|
| **1** | **Policy Evaluation** | Baseline metrics CSV (success rate, retry dist, error modes) | 🟢 ZERO | Metrics computed |
| **2** | **State Encoder + Intent Classifier** | sklearn model (behavior cloning) | 🟡 LOW | Accuracy ≥ 50% |
| **2-3** | **Imitation Learning** | Imitation policy trained on episodes | 🟡 LOW | Test accuracy ≥ 50% |
| **3-4** | **Offline RL** | Conservative Q-Learning policy | 🟡 MEDIUM | Trainer runs without divergence |
| **4** | **Shadow Mode** | Backend integration: LLM + ML run in parallel | 🟡 MEDIUM | Divergence < 10% |
| **5** | **Canary Rollout** | 10% traffic uses ML policy | 🔴 HIGH | Success rate ≥ LLM baseline |
| **5-6** | **Full Rollout** | 100% use ML (LLM fallback) | 🟡 MEDIUM | Monitoring stable for 3 days |

---

### Current Architecture (LLM-only)
```
User Message
    ↓
[LLM] Parse intent
    ↓
[LLM] Select tool (from allowlist)
    ↓
[Tool] Execute (deterministic)
    ↓
[LLM] Generate response text
```

### Target Architecture (ML + LLM Hybrid)
```
User Message
    ↓
[ML Classifier] Parse intent     ← supervised learning
    ↓
[ML Policy] Select tool           ← imitation + offline RL
    ↓
[Tool] Execute (same)             ← no change
    ↓
[LLM] Generate response text      ← same, but deterministic input
    ↓
[ML Classifier] Semantic tagging  ← supervised multi-label
```

---

## 📊 What Data We Have

✅ **50+ interaction episodes** per user (full decision sequences)  
✅ **2000+ tool calls** logged with status (success/timeout/error)  
✅ **Semantic index (WP7)** with AI-tagged categories + confidence  
✅ **State snapshots** (TM.json, PS.json, LO.json, GEN.json)  
✅ **Perfect for Imitation Learning v0** (baseline policy)

**Data Location**: `tmp/agentdatastorage_users_snapshot_20260210_105740/`

---

## 🔬 Phase 1: Policy Evaluation (Execute Today)

**Status**: ✅ Ready to run (0 risk, read-only analysis)

**What you get**:
```
labs/ml_lab/reports/
├── snapshot_summary.json       ← Episode counts, timestamp range
├── baseline_metrics.csv        ← Success rate, timeout %, retry dist
├── tool_analysis.csv           ← Per-tool success rate
├── policy_eval_baseline.html   ← Interactive report
└── PHASE_1_SUMMARY.md          ← Key findings + next steps
```

**Timeline**: 2-4 hours (mostly automated script runtime)

**Commands** (copy-paste):
```bash
cd "c:/AI memory/NewHope/OmniFlowBeta/labs/ml_lab"
python -m venv venv
.\venv\Scripts\Activate.ps1
pip install -r requirements.txt

python 2_policy_eval/00_snapshot_explorer.py
python 2_policy_eval/01_baseline_metrics.py
python 2_policy_eval/03_generate_report.py
```

**Success Criteria**:
- [ ] `success_rate` computed from episodes
- [ ] Per-tool metrics calculated
- [ ] HTML report opens without errors
- [ ] Baseline established for ML comparison

---

## 🚀 Integration Points (9 Components)

| Component | File | Function | Change Type |
|-----------|------|----------|-------------|
| **Intent Classifier** | `tool_call_handler/__init__.py` | `_pa_run_intention_step()` | Wrap with ML call |
| **State Encoder** | `tool_call_handler/__init__.py` | `_encode_user_state()` (NEW) | New function |
| **Tool Selector** | `tool_call_handler/__init__.py` | `main()` @ line 1250 | Branching logic |
| **Context Router** | `tool_call_handler/wp6/` | `decide_routing()` | Branching logic |
| **Semantic Classifier** | `tool_call_handler/__init__.py` | `_emit_semantic_audit()` | Branching logic |
| **Response Generator** | `tool_call_handler/__init__.py` | `main()` @ 1800 | Input pre-processing |
| **Reward Calculator** | `tool_call_handler/__init__.py` | `main()` | New artifact field |
| **Episode Recorder** | `tool_call_handler/__init__.py` | `main()` (end) | New artifact save |
| **Planner** | N/A (future) | N/A | Phase 4 |

See [ML_INTEGRATION_POINTS.md](https://github.com/dokuczacz/OmniFlowBeta/blob/main/tmp/ML_INTEGRATION_POINTS.md) for detailed code snippets.

---

## 🚨 Risk Mitigation

| Risk | Mitigation |
|------|-----------|
| **ML different from LLM** | 2-week shadow mode (log both, compare divergence) |
| **Reward shaped poorly** | Start simple (+1 success, -1 error), iterate |
| **State encoder too complex** | Start with interpretable features (count-based) |
| **Production latency spike** | Cache model outputs, fast LLM fallback |
| **Data leakage** | Lab uses read-only snapshots only |

**Rollback Plan**: Each phase has gates. LLM fallback always available via env var.

---

## ✅ Decision Checklist (Before Approval)

- [ ] **Timeline**: 6 weeks acceptable?
- [ ] **Risk profile**: Shadow mode + canary sufficient?
- [ ] **Resources**: Who owns Phase 1/2/3?
- [ ] **Success metric**: What's acceptable success rate threshold?
- [ ] **Fallback**: How to revert if ML underperforms?

---

### ✅ TODAY (Decision Phase)
1. Review `AI_DECISION_BLUEPRINT.md` (sections 1-5)
2. Approve timeline + risk gates
3. Assign Phase 1 owner

### ✅ THIS WEEK (Phase 1 Execution)
1. Setup `labs/ml_lab/` directory structure
2. Run Phase 1 scripts (policy evaluation)
3. Generate baseline metrics CSV
4. Review HTML report → decide Phase 2 go/no-go

### ✅ NEXT 2 WEEKS (Phase 2)
1. Manual intent labeling (20 samples)
2. Implement state encoder
3. Train imitation classifier
4. Evaluate accuracy on test set

### ✅ WEEK 3-6 (Phases 3-5)
1. Offline RL training
2. Shadow mode integration
3. Canary rollout (10% traffic)
4. Full rollout + monitoring

---

## 🎓 Key Concepts

- **Episode**: Full sequence `state → action → reward → new_state` (from interaction logs)
- **Imitation Learning**: Train ML to mimic historical "good" decisions (behavior cloning)
- **Offline RL**: Optimize policy using reward signals **without touching production**
- **Shadow Mode**: Run both LLM + ML, log divergence, use LLM for real (0 risk)
- **Planner**: Generate multi-step action sequences (goal decomposition)

---

## 📚 References

- **brainstorm**: [User's brain dump on ML/RL/Decision Intelligence](https://github.com/dokuczacz/OmniFlowBeta/blob/main/tmp/AI_DECISION_BLUEPRINT.md#1-rodzaje-danych-które-posiadamy)
- **Existing PA scope**: [`tmp/pa_scope_e2e_execplan.md`](https://github.com/dokuczacz/OmniFlowBeta/blob/main/tmp/pa_scope_e2e_execplan.md)
- **Tool handler refactor**: [REFACTOR_STATUS.md](https://github.com/dokuczacz/OmniFlowBeta/blob/main/REFACTOR_STATUS.md)

---

## 💬 Discussion Points

1. **Should Phase 1 start immediately** or wait for additional approvals?
2. **Who should own the ML model training** (Phase 2-3)?
3. **Acceptable success rate threshold** for canary rollout (e.g., ≥ 90% of LLM baseline)?
4. **Shadow mode duration**: 1 week or 2 weeks before canary?
5. **Fallback strategy**: Automatic or manual switch back to LLM?

---

## 📄 Files Modified (Commit)

This plan was committed as:
```
commit 17a4a955
docs: Add Decision Intelligence (ML/RL/Planner) transformation blueprint

- AI_DECISION_BLUEPRINT.md: Strategic plan
- ML_INTEGRATION_POINTS.md: Code integration guide
- EXECPLAN_PHASE_1.md: Policy Evaluation scripts
- INDEX_AND_QUICK_START.md: Quick reference
```

**Status**: ✅ Merged to `main`

---

**Labels**: `enhancement`, `machine-learning`, `planning`, `decision-intelligence`  
**Milestone**: Q1 2026  
**Assignees**: TBD (pending approval)

---

## ⚡ TL;DR

**What**: Replace LLM decision-making with ML/RL/Planner (LLM only for text generation)  
**Why**: Learn from 50+ episodes, optimize via RL, reduce cost 30%, faster inference  
**When**: 6 weeks (4 phases with gates)  
**Risk**: Low (offline training + shadow mode + canary)  
**Start**: Phase 1 ready today (2-4 hours, 0 risk)

**Approval needed**: Timeline + resources + success criteria

---

_Generated: 2026-02-10 by AI Agent (Claude Sonnet 4.5)_
